### PR TITLE
Rails image with standard entrypoint

### PR DESCRIPTION
**Pull request by thenonsequitur**
## Purpose

In the process of dockerizing our rails apps, we've added an entrypoint script that we've iterated on a decent amount. And we've been adding identical copies of this script to every rails app. It makes more sense to burn a base image with these entrypoint methods already built-in. So this PR creates a new base image for our rails apps.

## What does it do?

The Dockerfile:

* Starts with the ruby 2.3 base image
* Sets standard C.UTF-8 locale
* Adds the `docker-ssh-exec` binary (our current method of passing host ssh key to container)
* Adds a set of entrypoint helper scripts. Right now we only have two -- a script to clean up pids, and a mysql health check. But the idea is we can iterate to add more helper scripts or health checks.

## How will rails apps use it?

* For rails apps that want to use it, the base image to reference will be `FROM voxmedia/docker_base_images:rails-0.1`.
* Entrypoint helper scripts can be referenced from app entry points. Example of an app entrypoint using these scripts:

    #!/bin/sh
    set -e
    /opt/entrypoint/cleanup_pids.sh
    /opt/entrypoint/service_health_checks/mysql.sh
    exec docker-ssh-exec bundle exec "$@"

The `set -e` means the script will immediately exit if any command returns a non-zero status. Since the health checks will return non-zero on failure, simply executing the health check script is all that is needed.

Note that this last step, automatically running through `docker-ssh-exec`, means that when the container is started, it will copy the host's ssh key to the container, and it will stay there for the duration of the container's lifetime. Security-wise, this is not ideal, but it's also not particularly vulnerable to any attacks. If a local docker container is compromised, it most likely means the host machine itself is compromised, in which case we have bigger problems anyway.

The advantage of doing this is that we never have to think about which commands require `docker-ssh-exec` and it simplifies the dev experience quite a bit. I think this is definitely worth it.

## Launch plan

After merging, this image has to be published. The steps to publish are:

* Build locally. From within the `rails_entrypoint/` directory, `docker build .`, and note the image id it returns.
* Tag the image with `docker tag ID voxmedia/docker_base_images:rails_entrypoint-0.1`.
* Publish to docker hub with `docker push voxmedia/docker_base_images`.

This last step requires appropriate permissions on docker-hub.

### CODE REVIEW

File: rails/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,9 @@
+FROM voxmedia/docker_base_images:ruby_2.3-0.1
+
+ENV LC_ALL C.UTF-8
+
+ADD https://s3.amazonaws.com/voxmedia-docker-utils/docker-ssh-exec /usr/local/bin/docker-ssh-exec
+RUN chmod +x /usr/local/bin/docker-ssh-exec
+
+ADD docker_entrypoint.sh /usr/local/bin/docker_entrypoint.sh
+RUN chmod +x /usr/local/bin/docker_entrypoint.sh
```
Comment (thomsbg) [inline rails/Dockerfile]: Do we need to use the `ENTRYPOINT` instruction to actually set this as the default entrypoint?

File: rails/docker_entrypoint.sh [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+    else
+      sleep 1
+    fi
+  done
+
+  if [ $MYSQL_UP = "false" ]; then
+    echo "Unable to connect to database. Please make sure your ${MYSQL_HOST} container is running and healthy."
+    exit 1
+  fi
+fi
```
Comment (thomsbg) [inline rails/docker_entrypoint.sh]: This mysql check seems convenient, but feels like a weird precedent. Shouldn't all images based on this include their own dependency health checks as a best practice?

File: rails/Dockerfile (hunk 1 above)
Comment (thenonsequitur) [inline rails/Dockerfile]: I was leaving the ENTRYPOINT off intentionally to allow people to use the base image with a custom entrypoint (one that may reference this entrypoint) for other app-specific checks, but I guess I could also add it here. A custom entrypoint could still be used, a believe any subsequent ENTRYPOINT instruction will override this one.

File: rails/docker_entrypoint.sh (hunk 2 above)
Comment (thenonsequitur) [inline rails/docker_entrypoint.sh]: Hmm. I see your point, but I also really like the convenience. I wonder if tagging the image something else would help alleviate this concern? Maybe `voxmedia/docker_base_images:rails-mysql-entrypoint-0.1` or something equally verbose?

Comment (thenonsequitur) [inline rails/docker_entrypoint.sh]: An alternate idea would be to include a script with some standard health check functions. The apps would still be responsible for running health-checks, but they could be one-liners.

Comment (thomsbg) [inline rails/docker_entrypoint.sh]: Eh, no harm done by including it here, even if it is a little weird.

If I'm dreaming big, then I could imagine separating this check into a separate script. Then, this entrypoint or any other custom entrypoint could reference it. Even better would be a system that lets app developers drop health check scripts into a certain directory or file, and have those automatically invoked by this entrypoint.

Comment (thenonsequitur) [inline rails/docker_entrypoint.sh]: Yeah I think as a first pass I'll just separate this out into a different script that apps can reference, I think that's an improvement and does make the structure less weird. I might pass on a more substantial system for dropping in health checks, maybe a task for a future time.

**Pull request merged**